File: Alcoholtracker/Services/WaterLog.swift
import Foundation

// MARK: - WaterLog
//
// Per-day water glass counter, keyed by logical day (06:00 boundary).
// Stored in UserDefaults; feeds the hydration widget and the hangover
// prediction. nil means the user never logged water that day, which lets
// callers fall back to the old estimation heuristic instead of assuming zero.

enum WaterLog {

    static let glassML = 250.0

    private static let storageKey = "waterLog_v1"

    private static var storage: [String: Int] {
        get { UserDefaults.standard.dictionary(forKey: storageKey) as? [String: Int] ?? [:] }
        set { UserDefaults.standard.set(newValue, forKey: storageKey) }
    }

    // Key for a calendar day (any time-of-day component is stripped).
    private static func key(forDay day: Date) -> String {
        let comps = Calendar.current.dateComponents([.year, .month, .day], from: day)
        return String(format: "%04d-%02d-%02d", comps.year ?? 0, comps.month ?? 0, comps.day ?? 0)
    }

    private static var todayKey: String {
        key(forDay: Calendar.current.logicalDay(for: Date()))
    }

    // MARK: Reading

    /// Glasses logged for a specific calendar day; nil if nothing was ever logged.
    static func loggedGlasses(forDay day: Date) -> Int? {
        storage[key(forDay: Calendar.current.startOfDay(for: day))]
    }

    static func glassesToday() -> Int {
        storage[todayKey] ?? 0
    }

    // MARK: Writing (always against the current logical day)

    static func addGlassToday() {
        var s = storage
        s[todayKey, default: 0] += 1
        storage = s
    }

    static func removeGlassToday() {
        var s = storage
        s[todayKey] = max(0, (s[todayKey] ?? 0) - 1)
        storage = s
    }

    // MARK: Account backup

    /// Every logged day, keyed "YYYY-MM-DD". Used by HistorySyncService to back
    /// up the water log to the signed-in account.
    static var allEntries: [String: Int] { storage }

    /// Merges restored entries into the local log, keeping the higher count per
    /// day so a backup never lowers a freshly logged value.
    static func merge(_ entries: [String: Int]) {
        guard !entries.isEmpty else { return }
        var s = storage
        for (day, glasses) in entries {
            s[day] = max(s[day] ?? 0, glasses)
        }
        storage = s
    }
}

File: android/bac/src/main/kotlin/de/tipau/promille/bac/WaterLog.kt
package de.tipau.promille.bac

import java.time.LocalDate
import java.time.ZoneId
import java.util.Locale

/**
 * Glasses of water the user logged, keyed by LOGICAL day. Mirrors
 * Services/WaterLog.swift, including the key format: HistorySyncService backs the
 * table up to Supabase, so an Android row has to be readable by iOS and back.
 *
 * Storage is injected so this stays a plain JVM class.
 */
class WaterLog(private val store: Store) {

    /** Persists the whole table at once, exactly like the iOS JSON blob does. */
    interface Store {
        fun load(): Map<String, Int>
        fun save(entries: Map<String, Int>)
    }

    val allEntries: Map<String, Int> get() = store.load()

    /**
     * Null means "never logged", which lets the caller fall back to the estimation
     * heuristic instead of reading a silent zero as a real zero.
     */
    fun loggedGlasses(day: LocalDate): Int? = store.load()[key(day)]

    fun loggedGlasses(epochSeconds: Long, zone: ZoneId = ZoneId.systemDefault()): Int? =
        loggedGlasses(LogicalDay.dateOf(epochSeconds, zone))

    fun glassesToday(nowEpochSeconds: Long, zone: ZoneId = ZoneId.systemDefault()): Int =
        loggedGlasses(nowEpochSeconds, zone) ?: 0

    fun addGlassToday(nowEpochSeconds: Long, zone: ZoneId = ZoneId.systemDefault()) {
        val k = key(LogicalDay.dateOf(nowEpochSeconds, zone))
        val entries = store.load().toMutableMap()
        entries[k] = (entries[k] ?: 0) + 1
        store.save(entries)
    }

    fun removeGlassToday(nowEpochSeconds: Long, zone: ZoneId = ZoneId.systemDefault()) {
        val k = key(LogicalDay.dateOf(nowEpochSeconds, zone))
        val entries = store.load().toMutableMap()
        val next = (entries[k] ?: 0) - 1
        entries[k] = if (next < 0) 0 else next
        store.save(entries)
    }

    /**
     * Keeps the HIGHER count per day, so restoring a backup can never lower a
     * value the user has already logged today on this device.
     */
    fun merge(other: Map<String, Int>) {
        val entries = store.load().toMutableMap()
        for ((day, count) in other) {
            val existing = entries[day] ?: 0
            if (count > existing) entries[day] = count
        }
        store.save(entries)
    }

    fun clear() = store.save(emptyMap())

    companion object {
        const val GLASS_ML = 250.0

        /** Key of the backing blob, shared with iOS. */
        const val STORAGE_KEY = "waterLog_v1"

        /** Wire format of a day key. Locale.ROOT so digits never localise. */
        fun key(day: LocalDate): String =
            String.format(Locale.ROOT, "%04d-%02d-%02d", day.year, day.monthValue, day.dayOfMonth)

        fun inMemory(): WaterLog = WaterLog(object : Store {
            private var entries: Map<String, Int> = emptyMap()
            override fun load(): Map<String, Int> = entries
            override fun save(entries: Map<String, Int>) { this.entries = entries }
        })

        /** Never records anything. For previews and pure math. */
        fun disabled(): WaterLog = WaterLog(object : Store {
            override fun load(): Map<String, Int> = emptyMap()
            override fun save(entries: Map<String, Int>) = Unit
        })
    }
}
